File: 프로그래머스/3/43164. 여행경로/여행경로.py
from collections import defaultdict

def solution(tickets):
    t_dict = defaultdict(list)
    # result = sorted(tickets, key = lambda x: ((x[0], x[1])))

    for s, e in tickets:  # tickets 원소들을 출발 공항을 key, 도착 공항을 value로 저장
        t_dict[s].append(e)
    
    for t_key in t_dict.keys():  # 도착 공항 리스트를 내림차순으로 정렬 -> pop() 해서 사용할 예정이라 
        t_dict[t_key].sort(reverse = True)
    
    answer = []
    path = ["ICN"]  # 출발지 "ICN" 
    
    while path:  # DFS 실행
        now = path[-1]
        
        if now not in t_dict or len(t_dict[now]) == 0:  # 이어지는 공항이 없는 경우와 이미 항공권을 사용한 경우
            answer.append(path.pop())  # 위 조건을 만족할 경우 정답인 경로
        else:  # now에서 출발하는 항공편 path에 추가
            path.append(t_dict[now].pop())
        print(path)
    return answer[::-1]  # 내림차순으로 정렬했기 때문에 뒤집어서 return 
    

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    public String[] solution(String[][] tickets) {
        Arrays.sort(tickets, (o1, o2) -> o1[1].compareTo(o2[1]));
        boolean[] visited = new boolean[tickets.length];
        
        List<String> path = new ArrayList<>();
        path.add("ICN");

        List<String> result = dfs("ICN", tickets, path, visited);
        
        return result.toArray(new String[0]);
    }

    
    public List<String> dfs(String cur, String[][] tickets, List<String> path, boolean[] visited) {
        if (path.size() == tickets.length + 1) {
            return new ArrayList<>(path);
        }

        for (int i = 0; i < tickets.length; i++) {
            if (!visited[i] && tickets[i][0].equals(cur)) {
                visited[i] = true;
                path.add(tickets[i][1]);
                
                List<String> result = dfs(tickets[i][1], tickets, path, visited);
                
                if (result != null) return result;
                
                path.remove(path.size() - 1);
                visited[i] = false;
            }
        }

        return null;
    }
}
